### PR TITLE
Lisp url args

### LINKED ISSUE #2397: Allow lisp:// URL to receive arguments again
As discussed in https://github.com/atlas-engineer/nyxt/pull/2377#issuecomment-1157741439, `lisp://` URLs no longer take arguments and this can be problematic in some cases:

- In delete-bookmark, we have to generate 1 closure per bookmark which is not ideal performance-wise.
- What if we get some data from JS, how do we communicate it to the `lisp://` URL closure?

It's easy to extend the `lisp` scheme to accept argument (for instance as URL query arguments, just like `nyxt://` URLs).

Then the question that remains is how we refer to the lambda-list.

Currently we have this syntax:

```lisp
(ps:ps (nyxt/ps:lisp-eval
        (:title "describe-any")
        (nyxt::describe-any (princ-to-string symbol))))
```

I guess we could have this instead?

```lisp
(ps:ps (nyxt/ps:lisp-eval (&key key-arg1 key-arg2)
        (:title "describe-any")
        (log:info key-arg1 key-arg2)
        (nyxt::describe-any (princ-to-string symbol))))
```

But then we need to change all the `lisp-eval` calls.

Thoughts? @aartaka

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This adds the ability to provide arbitrary keyword arguments to `lisp:` URLs and parse those in the callbacks. In addition to that there's also:
- `lisp-url` to construct `lisp:`-schemed URLs.
- Ability to only generate a callback once, while using it several times (`:id` argument to both `lisp-eval` and `lisp-url`).

Fixes #2397.

# Discussion

- This patch requires the caller to specify arguments twice: in the `nyxt/ps:lisp-eval` ergument list, as keywords, and in the body of `lisp-eval` (and `lisp-url`), as lambda with `(&key arg1 arg2)`. This is duplicative, but I haven't found a reasonable way to generate lambda lists from the keyword lists, given that the `*package*` is almost always bound to `nyxt-user` and thus does not allow to generate the package-relevant argument names. 
  - For example: using `href` in the body of `lisp-eval` in `list-bookmarks` refers to the `nyxt/list-bookmarks-mode::href`, while `(intern (symbol-name :HREF) *package*)` in `lisp-eval` results in `nyxt-user:href`, breaking the whole thing with symbol undefined error.
- `lisp-url` generates URLs with the question mark at the end, to ease argument concatenation on the JS-side. Should we do that, though? Given that we concatenate `title` to it, having a guarantee of URL ending at either `?` or `&` is a virtue, but it doesn't look right.
- This makes syntax of `lisp:` URLs slightly backwards-incompatible, but they were not saved to history anyway, so it shouldn't break many workflows.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [X] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
    - Maybe add some `lisp-url` generation tests?
  - [X] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: source/urls.lisp [hunk 1]
```diff
@@ -399,31 +399,69 @@ guarantee of the same result."
...
+(ps:defpsmacro nyxt/ps::lisp-eval ((&rest args &key (buffer '(nyxt:current-buffer)) title id callback
+                                          &allow-other-keys)
+                                   &body body)
   "Request the lisp: URL and invoke CALLBACK when there's a successful result.
+
+ID is either caller-provided or auto-generated.
+- If it's caller-provided, this ID/CALLBACK is reusable with different
+  ARGS.
+- If it's auto-generated, it's not accessible to the caller and thus not
+  reusable.
```
Comment (Ambrevar) [inline source/urls.lisp:432]: Can you explain why the user would want access to the ID?  Maybe with an example?

File: source/urls.lisp [hunk 2]
```diff
@@ -399,31 +399,69 @@ guarantee of the same result."
...
-     (let ((request (fetch url
-                           (ps:create :mode "no-cors"))))
-       (when ,callback
-         (chain request
-                (then (lambda (response)
-                        (when (@ response ok)
-                          (chain response (json)))))
-                (then ,callback))))))
+  (let ((irregular-args (alexandria:remove-from-plist
+                         args :id :buffer :callback :title)))
```
Comment (Ambrevar) [inline source/urls.lisp:442]: This is awkward.  What about making `args` a required argument or a keyword argument, available as a sublist?

Example: `(:buffer BUFFER1 :args '(:foo 17 :bar 18))`

File: source/urls.lisp [hunk 3]
```diff
@@ -399,31 +399,69 @@ guarantee of the same result."
...
+    `(let ((url (ps:lisp (let ((request-id (or ,id (string (gensym "")))))
+                           (str:concat
+                            (lisp-url
+                             :id request-id
+                             :buffer ,buffer
+                             :callback ,(if (and (sera:single body)
+                                                 (member (first (first body)) '(lambda function)))
+                                            (first body)
+                                            `(lambda ()
+                                               ,@body))
```
Comment (Ambrevar) [inline source/urls.lisp:452]: Style: Maybe lambda on just one line?

**Comment by Ambrevar** [comment]
About the duplicated arguments, what about using arbitrary keyword argument names?
```lisp
(defun foo (&key ((:href my-href)))
   (log:info my-href))
```
Alternatively, you could generate the lambda-list, no?

File: source/urls.lisp (hunk 1 above)
Comment (aartaka) [reply source/urls.lisp:432]: Reusing the same callback several times. For example, you can pre-make and re-use the callback, as in
``` lisp
(defvar id (nyxt:new-id))
;; Create a callback
(nyxt/ps:lisp-eval
 (:title "delete bookmark" :id id)
 (lambda (&key href)
   (nyxt/bookmark-mode:delete-bookmark href)))
;; Use it
(nyxt/ps:lisp-eval
 (:id id :href "https://github.com/"))
(nyxt/ps:lisp-eval
 (:id id :href "https://gitlab.com/"))
```

Comment (Ambrevar) [reply source/urls.lisp:432]: If there is no other use case, then maybe replace the ID param with a boolean, like `reuse-p`, and have it remember / reuse the same ID automatically.

Currently this `:id` param feels too low level and leaves room for too many mistakes.

Comment (aartaka) [reply source/urls.lisp:432]: What can help there is separation of `define` and `funcall` stages for `lisp-eval`. Right now the mental distance JS world is located at makes us use this `define-and-funcall-sometime` construction that this design is inspired by :)

File: source/urls.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/urls.lisp:442]: You can parse arbitrary arglists with `destructuring-bind`.

Comment (aartaka) [reply source/urls.lisp:442]: > You can parse arbitrary arglists with `destructuring-bind`.

Fair commentary, yes. Still, I don't see much hurt in inlining the arguments there. Especially if we talk of separating the "definition" and the "funcall" stages of the `lisp-eval`.

**Comment by Ambrevar** [comment]
Can you give a full example of how to reproduce this issue?

Internal page macros don't seem to be setting the package, do they?

File: source/urls.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/urls.lisp:442]: Not much hurt, at least for now, but it's more future-proof.

File: source/urls.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/urls.lisp:432]: Sorry, can you rephrase?  Maybe give an example?

**Comment by aartaka** [comment]
> Can you give a full example of how to reproduce this issue?

Basically parse any keyword inside the internal page body, and you'll get a `nyxt-user`-interned symbol.

> Internal page macros don't seem to be setting the package, do they?

Internal pages don't, but start.lisp does:
https://github.com/atlas-engineer/nyxt/blob/540022904ff82915dc3006a92d8588c77e0f843c/source/start.lisp#L642

**Comment by aartaka** [comment]
@Ambrevar, Ive just pushed a refactoring of our `lisp-url` handling. See ddfb8fe9df1ba6c20d12a58be36d3592314653c9 for details, but the highlight is that we now have separation of `lisp-url` definition and call, and a convenient `lisp-url-flet` macro to save us dozens of lines!

File: source/urls.lisp [hunk 4]
```diff
@@ -400,31 +400,118 @@ guarantee of the same result."
                     (t (error "Cannot display evaluation result"))))))))))
   :local-p t)
 
-(ps:defpsmacro nyxt/ps::lisp-eval ((&key (buffer '(nyxt:current-buffer)) title callback) &body form)
+(export-always 'lisp-url)
```
Comment (Ambrevar) [inline source/urls.lisp:403]: Shouldn't it be unexported if `lisp-url-flet` and `lisp-call` are the high level API points?

Comment (aartaka) [reply source/urls.lisp:403]: Hmmmmmm... I'm not sure. I can't imagine a solid use-case for it that `lisp-url-flet` an `lisp-call` woudn't cover, but I certainly am no omniscient, so...

I'll remove it.

**Comment by aartaka** [comment]
I'm still not sure about the benefits of making arguments a dedicated argument, especially given the fact that we're already in the foggy waters of Parenscript compile-time JS generation. I don't want to add any more quasi-quote madness to it (like, do you need to quote args? Do you want to provided a quoted list or an unquoted one? Do you want to pre-evaluate thing? Too much non-obvious questions there.)

**Comment by Ambrevar** [comment]
Anything blocking here?

File: source/urls.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/urls.lisp]: Friendly ping :)

**Comment by Ambrevar** [comment]
@aartaka Have you tested the callbacks?

**Comment by aartaka** [comment]
> @aartaka Have you tested the callbacks?

No, I haven't, because I'm still implying they don't work.

**Comment by aartaka** [comment]
> Anything blocking here?

Nothing but deciding whether we want the `:args '(:arg1 4 :arg2 ,(whatever))` or `:arg1 4 :arg2 (whatever)` syntax :)

**Comment by Ambrevar** [comment]
Let's wrap this up after 3-pre-release-2.

**Comment by aartaka** [comment]
I've updated this changeset to have `:args` as the separate argument to `lisp-(call|eval)`, but haven't tested it yet.

**Comment by aartaka** [comment]
It works and is ready for review!

File: source/urls.lisp [hunk 5]
```diff
@@ -417,30 +417,112 @@ guarantee of the same result."
...
+      (setf (gethash id (nyxt::lisp-url-callbacks buffer)) callback)))
+  (format nil "lisp://~a?~:[~*~;title=~a&~]" id
+          title (when title
+                  (quri:url-encode title))))
+
+
+(ps:defpsmacro nyxt/ps::lisp-call (id &key title (buffer (current-buffer)) args)
+  "Call the ID-bound function on the Lisp side.
+
+Return a Promise succeeding after the code runs on the Lisp side.
```
Comment (Ambrevar) [inline source/urls.lisp:443]: "promise fulfilled after"

Comment (Ambrevar) [reply source/urls.lisp:443]: It's a JS promise, right?

File: source/urls.lisp [hunk 6]
```diff
@@ -417,30 +417,112 @@ guarantee of the same result."
...
+    ((hello (&key name) (nyxt:echo \"Hello, ~a!\" name)))
+  (nyxt:ffi-buffer-evaluate-javascript (nyxt:current-buffer) (hello :name \"Stranger\"))
+  ;; or equivalent
+  (nyxt:ffi-buffer-evaluate-javascript
+   (nyxt:current-buffer)
+   (ps:ps (nyxt/ps:lisp-call hello :buffer (nyxt:current-buffer) :args '(:name \"Stranger\"))))))
+
+BEWARE: While `lisp-url's defined by this macro stay defined for a little while
+longer than this macro lasts, it's not recommended to use them outside the
+scope."
```
Comment (Ambrevar) [inline source/urls.lisp:477]: Is it worth mentioning this at all?

**Review by Ambrevar — COMMENTED** [review]
Looks good to me!

**Comment by aartaka** [comment]
Fixed documentation and rebased against master. Merging?

**Comment by aadcg** [comment]
I believe what you have pushed is the same set of commits.

**Comment by aartaka** [comment]
Yes, now it should be two commits. Although I have to test whether I missed anything during rebase...

**Comment by aadcg** [comment]
Thanks @aartaka!
